### PR TITLE
Add `advice` to `EvaluationError`s

### PR DESCRIPTION
**Pull request by khieta**
## Description of changes

* Refactored `EvaluationError` into the following form, which matches our current `ValidationError` structure. The `advice` field contains an optional string with additional help information. If we were using [miette](https://docs.rs/miette/latest/miette/), this field would be labeled as `#[help]`. But for this PR, I'm considering updating to miette out of scope (see issue #182). This new structure should make supporting location information simpler in the future.
```
pub enum EvaluationError {
  advice: Option<String>,
  error_kind: EvaluationErrorKind.
}
```
* Added advice in 4 error cases that seem likely:
  * Using `in` instead of `.contains()`
     `3 in [34, -2, 7]` --> "type error: expected (entity of type any_entity_type), got long. `in` is for checking the entity hierarchy, use `.contains()` to test set membership"
  * Using `in` instead of `has`
    `"foo" in { "foo": 2, "bar": true }` --> "type error: expected (entity of type any_entity_type), got string. `in` is for checking the entity hierarchy, use `has` to test if a record has a key"
  * Using an IP-related function on a string instead of an IP value
    `"127.0.0.1".isIpv4()` --> "type error: expected ipaddr, got string. Maybe you forgot to apply the `ip` constructor?"
  * Using a decimal-related function on a string instead of a decimal value
    `"1.23".lessThan("2.34")` --> "type error: expected decimal, got string. Maybe you forgot to apply the `decimal` constructor?"

## Issue #, if available

#177

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)

I confirm that this PR (choose one, and delete the other options):

- [X] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by cdisselkoen — APPROVED** [review]
To your questions in the description: 

Yes, I think it's a good idea to make that structural change to `EvaluationError`.

No, I don't think it has to be in scope for this PR (but if it isn't, let's make a separate issue for the backlog?)

Something I do think we could do in this PR is make the same change for `decimal` that we did for `ip` -- i.e., a hint for the case `"23.57".greaterThan(...)`

**Comment by khieta** [comment]
Finished the first pass of refactoring! The diff is larger now, but the only important changes are in `evaluator/err.rs`, `extensions/ipaddr.rs`, `extensions/decimal.rs`, and `evaluator.rs` L398-411.

File: cedar-policy-core/src/ast/extension.rs [hunk 1]
```diff
@@ -158,11 +158,11 @@ impl ExtensionFunction {
...
-                    Err(evaluator::EvaluationError::WrongNumArguments {
-                        function_name: name.clone(),
-                        expected: 0,
-                        actual: args.len(),
-                    })
+                    Err(evaluator::EvaluationError::wrong_num_arguments(
+                        name.clone(),
+                        0,
+                        args.len(),
+                    ))
```
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/extension.rs:165]: side note that it's a shame this has to become less readable like this, I wish Rust had named arguments to functions

File: cedar-policy-core/src/evaluator.rs [hunk 2]
```diff
@@ -3777,6 +3648,16 @@ pub mod test {
         );
     }
 
+    fn assert_restricted_expression_error(v: Result<PartialValue>) {
+        match v {
+            Err(e) => assert!(matches!(
```
Comment (cdisselkoen) [inline cedar-policy-core/src/evaluator.rs:3653]: `assert_matches!` gives nicer errors than `assert!(matches!(` which just will say "hey this was false and expected true"

File: cedar-policy-core/src/extensions/decimal.rs [hunk 3]
```diff
@@ -290,17 +301,20 @@ mod tests {
...
+                    println!("{msg}");
+                    assert_eq!(
+                        *extension_name,
+                        Name::parse_unqualified_name("decimal")
+                            .expect("should be a valid identifier")
+                    )
+                }
+                _ => panic!("Expected an decimal ExtensionErr, got {:?}", e),
+            },
             Ok(_) => panic!("Expected an decimal ExtensionErr, got Ok"),
```
Comment (cdisselkoen) [inline cedar-policy-core/src/extensions/decimal.rs:318]:
```suggestion
                _ => panic!("Expected a decimal ExtensionErr, got {:?}", e),
            },
            Ok(_) => panic!("Expected a decimal ExtensionErr, got Ok"),
```

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy-core/src/evaluator.rs (hunk 2 above)
Comment (khieta) [reply cedar-policy-core/src/evaluator.rs:3653]: `assert_matches` is a nightly feature, so to enable it we'd have to add `#![feature(assert_matches)]` to the crate -- are we ok with adding this crate-level attribute?

**Review by john-h-kastner-aws — APPROVED** [review]
Looks great.

**Pull request merged**